File: dev/optimising/CUBA/threshold-call.cpp
/*
 * Calling with netsize = 4000 and repeats = 100000 takes about 11s compared to
 * Python's 92s of which NumPy's 40s so this looks to be a candidate for optimisation
 * 
 */

#include<cstdlib>
#include<ctime>
#include<iostream>
#include<list>

const int netsize = 4000; // 4000 for comparison with cuba-profile.txt
const int repeats = 50000; // 400000 for comparison with cuba-profile.txt

using namespace std;

double timeforfiringfraction(double *V, double Vt)
{
	clock_t start = clock();
	list<int> l;
	for(int i=0;i<repeats;i++)
	{
		l.clear();
		for(int j=0;j<netsize;j++)
		{
			if(V[j]>Vt) l.push_back(j);
		}
	}
	return (clock()-start)*1./CLOCKS_PER_SEC;
}

// This is the method used by the .nonzero() method of NumPy
double timeforfiringfraction2(double *V, double Vt)
{
	clock_t start = clock();
	for(int i=0;i<repeats;i++)
	{
		int count = 0;
		for(int j=0;j<netsize;j++)
			if(V[j]>Vt) count++;
		int *l = new int [count];
		int *lptr = l;
		for(int j=0;j<netsize;j++)
		{
			if(V[j]>Vt) *lptr++ = j;
		}
		delete l;
	}
	return (clock()-start)*1./CLOCKS_PER_SEC;
}

int main(void)
{
	double V[netsize];
	srand((unsigned int)time(NULL));
	for(int i=0;i<netsize;i++)
		V[i]=double(rand())/(double)RAND_MAX;
	int netrateinc = 1;
	for(int netrate=1;netrate<=400;netrate+=netrateinc)
	{
		if(10<=netrate & netrate<100)
			netrateinc = 10;
		if(100<=netrate)
			netrateinc = 50;
		double Vt = 1 - (double)netrate/4000.;
		cout << netrate << ",  " << timeforfiringfraction(V,Vt) << ", " << timeforfiringfraction2(V,Vt) << endl;
	}
	return 0;
} 

File: dev/optimising/CUBA/threshold-call.py
"""

This appears to be an accurate model of the time spent in the threshold
calling mechanism, and so can be used as a basis for optimising.

With netsize = 4000 and repeats = 400000 as in the cuba-profile.txt example,
using original method gives 40s in nonzero method of numpy.ndarray, which
accords with what happens in the CUBA profile, and 11s in
FakeThreshold.__call__ which also accords.

"""

from numpy import *
import c_profile
import time

## parameters for comparison with cuba-profile.txt
#netsize = 4000
#repeats = 400000
# parameters for quicker runs to try out optimisations
# original method: 4.768s in nonzero, 1.366 in __call__
netsize=4000
repeats=50000


class FakeNeuronGroup(object):
    pass


class FakeThreshold(object):
    # this is the original method
    def __call__(self, P):
        return ((P.S[0]>self.Vt).nonzero())[0]

V=random.uniform(size=netsize)
Vt=1-0.0005

P=FakeNeuronGroup()
for i in range(40):
    exec 'P.a'+str(i)+'=1'
P.S=zeros((3, netsize))
P.S[0]=V

T=FakeThreshold()
#T.Vt = Vt
for i in range(40):
    exec 'T.a'+str(i)+'=1'

def main():
    for i in xrange(repeats):
        spikes=T(P)
c_profile.run('main()')

#def timeforfiringfraction(Vt):
#    global T
#    T.Vt = Vt
#    start = time.time()
#    for i in xrange(repeats):
#        spikes = T(P)
#    return time.time()-start
#
#def main():
#    for rate in [1,5,10,50,100,400,4000]:
#        Vt = 1 - float(rate)/4000.
#        print rate, ",", timeforfiringfraction(Vt)
#        
#c_profile.run('main()')
